### PR TITLE
Refactor for RAII

### PR DESCRIPTION
**Pull request by guolinke**
Refactor principle:
1. use ```std::unique_ptr``` for pointer, and ```std::vector``` for arrays
2. only warp ```std::unique_ptr``` for the pointers that need to take care for ownership
3. For the pointer that passed from outside and just use in current class, still use raw pointers. Because these spaces are managed by other classes. 
4. Not hurt to speed

### CODE REVIEW

File: include/LightGBM/bin.h [hunk 1]
```diff
@@ -231,7 +231,7 @@ class Bin {
   * \param out Output Result
   */
   virtual void ConstructHistogram(
-    data_size_t* data_indices, data_size_t num_data,
+    const data_size_t* data_indices, data_size_t num_data,
```
Comment (xuehui1991) [inline include/LightGBM/bin.h]: I wonder can use "std::vector" instead of "data_size_t *" here?

Comment (xuehui1991) [inline include/LightGBM/bin.h]: I wonder can use "std::vector" instead of "data_size_t *" here?

Comment (xuehui1991) [inline include/LightGBM/bin.h]: I wonder can use "std::vector" instead of "data_size_t *" here?

File: include/LightGBM/utils/common.h [hunk 2]
```diff
@@ -359,29 +360,15 @@ inline void Softmax(std::vector<double>* p_rec) {
...
-  } else {
-    std::sort(arr.begin(), arr.end(), [](const std::pair<T1, T2>& a, const std::pair<T1, T2>& b) {
-      return a.first > b.first;
-    });
-  }
-  for (size_t i = start; i < arr.size(); ++i) {
-    keys[i] = arr[i].first;
-    values[i] = arr[i].second;
+template<typename T>
+std::vector<const T*> ConstPtrInVectorWarpper(const std::vector<std::unique_ptr<T>>& input) {
```
Comment (xuehui1991) [inline include/LightGBM/utils/common.h]: ConstPtrIntoVectorWarpper?

File: src/application/application.cpp [hunk 3]
```diff
@@ -138,37 +119,40 @@ void Application::LoadData() {
...
+      train_metric_.push_back(std::move(metric));
     }
   }
   // Add validation data, if it exists
   for (size_t i = 0; i < config_.io_config.valid_data_filenames.size(); ++i) {
     // add
-    valid_datas_.push_back(dataset_loader_->LoadFromFileAlignWithOtherDataset(config_.io_config.valid_data_filenames[i].c_str(),
-      train_data_));
+    auto new_dataset = std::unique_ptr<Dataset>(
+      dataset_loader.LoadFromFileAlignWithOtherDataset(
```
Comment (xuehui1991) [inline src/application/application.cpp]: the name "LoadFromFileAlignWithOtherDataset" still not change....

File: src/boosting/dart.hpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+#include "gbdt.h"
+
+#include <cstdio>
+#include <vector>
+#include <string>
+#include <fstream>
+
+namespace LightGBM {
+/*!
+* \brief DART algorithm implementation. including Training, prediction, bagging.
```
Comment (xuehui1991) [inline src/boosting/dart.hpp]: training?

File: src/boosting/dart.hpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  * \return training score
+  */
+  const score_t* GetTrainingScore(data_size_t* out_len) override {
+    DroppingTrees();
+    *out_len = train_score_updater_->num_data() * num_class_;
+    return train_score_updater_->score();
+  }
+  /*!
+  * \brief Serialize models by string
+  * \return String output of tranined model
```
Comment (xuehui1991) [inline src/boosting/dart.hpp]: trained

File: src/boosting/dart.hpp [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    GBDT::Init(config, train_data, object_function, training_metrics);
+    gbdt_config_ = config;
+    drop_rate_ = gbdt_config_->drop_rate;
+    shrinkage_rate_ = 1.0;
+    random_for_drop_ = Random(gbdt_config_->drop_seed);
+  }
+  /*!
+  * \brief one training iteration
+  */
+  bool TrainOneIter(const score_t* gradient, const score_t* hessian, bool is_eval) override {
```
Comment (xuehui1991) [inline src/boosting/dart.hpp]: mark here....similar code...

File: src/io/dataset_loader.cpp [hunk 7]
```diff
@@ -263,31 +259,30 @@ Dataset* DatasetLoader::LoadFromBinFile(const char* bin_filename, int rank, int
 
   // buffer to read binary file
   size_t buffer_size = 16 * 1024 * 1024;
```
Comment (xuehui1991) [inline src/io/dataset_loader.cpp]: why no set buffer_size  as a const variable?

File: src/treelearner/serial_tree_learner.cpp [hunk 8]
```diff
@@ -103,20 +78,20 @@ void SerialTreeLearner::Init(const Dataset* train_data) {
...
 
-  is_feature_used_ = new bool[num_features_];
+  is_feature_used_.resize(num_features_);
 
   // initialize ordered gradients and hessians
-  ordered_gradients_ = new score_t[num_data_];
-  ordered_hessians_ = new score_t[num_data_];
+  ordered_gradients_.resize(num_data_);
+  ordered_hessians_.resize(num_data_);
   // if has ordered bin, need allocate a buffer to fast split
```
Comment (xuehui1991) [inline src/treelearner/serial_tree_learner.cpp]: // if has ordered bin, need to allocate a buffer to fast split

File: include/LightGBM/bin.h (hunk 1 above)
Comment (guolinke) [inline include/LightGBM/bin.h]: I try to avoid changing too much code. so just live with it ~

File: include/LightGBM/utils/common.h (hunk 2 above)
Comment (guolinke) [inline include/LightGBM/utils/common.h]: You means need better name?

File: src/application/application.cpp (hunk 3 above)
Comment (guolinke) [inline src/application/application.cpp]: Just give me a better name= =...

File: include/LightGBM/utils/common.h (hunk 2 above)
Comment (wxchan) [inline include/LightGBM/utils/common.h]: Warpper is a misspell -> Wrapper.

**Pull request merged**